apps/sel4test-tests/src/tests/cache.c

```diff
@@ -253,3 +254,78 @@ DEFINE_TEST(CACHEFLUSH0003, "Test that cache maintenance can be done on large pa
             config_set(CONFIG_HAVE_CACHE))
 
 #endif
+
+
+static int test_page_uncached_after_retype(env_t env)
+{
+    seL4_CPtr frame, frame2, untyped;
+    uintptr_t vstart, vstart2;
+    volatile uint32_t *ptr, *ptr2;
+    vka_t *vka;
+    int error;
+
+    vka = &env->vka;
+
+    // Create 2 vspace reservations because when we revoke the untyped the state gets out of sync.
+    void *vaddr;
+    void *vaddr2;
+    reservation_t reservation, reservation2;
+
+    reservation = vspace_reserve_range(&env->vspace,
+                                       PAGE_SIZE_4K, seL4_AllRights, 0, &vaddr);
+    assert(reservation.res);
+    reservation2 = vspace_reserve_range(&env->vspace,
+                                        PAGE_SIZE_4K, seL4_AllRights, 0, &vaddr2);
+    assert(reservation2.res);
+
+    vstart = (uintptr_t)vaddr;
+    assert(IS_ALIGNED(vstart, seL4_PageBits));
+    vstart2 = (uintptr_t)vaddr2;
+    assert(IS_ALIGNED(vstart2, seL4_PageBits));
+
+    ptr = (volatile uint32_t *)vstart;
+    ptr2 = (volatile uint32_t *)vstart2;
+
+    /* Get an untyped */
+    untyped = vka_alloc_untyped_leaky(vka, PAGE_BITS_4K);
+    test_assert(untyped != seL4_CapNull);
+    cspacepath_t src_path;
+    vka_cspace_make_path(vka, untyped, &src_path);
+
+    frame = get_free_slot(env);
+    test_assert(frame != seL4_CapNull);
+    cspacepath_t dest_path;
+    vka_cspace_make_path(vka, frame, &dest_path);
+    /* Create a frame from the untyped */
+    error = seL4_Untyped_Retype(untyped, seL4_ARCH_4KPage, PAGE_BITS_4K, dest_path.root, dest_path.dest,
+                                dest_path.destDepth, dest_path.offset, 1);
+    test_error_eq(error, seL4_NoError);
+
+    /* map in the without cacheability */
+    error = vspace_map_pages_at_vaddr(&env->vspace, &frame, NULL, vaddr, 1, seL4_PageBits, reservation);
+    test_error_eq(error, seL4_NoError);
+
+    /* Write some data directly to RAM as it's uncached */
+    *ptr = 0xC0FFEE;
+
+    /* Revoke the untyped. This deletes the frame. */
+    vka_cnode_revoke(&src_path);
+
+    /* Create the frame with the same untyped. The kernel guarantees that the contents have been zeroed. */
+    error = seL4_Untyped_Retype(untyped, seL4_ARCH_4KPage, PAGE_BITS_4K, dest_path.root, dest_path.dest,
+                                dest_path.destDepth, dest_path.offset, 1);
+    test_error_eq(error, seL4_NoError);
+
+
+    /* map in the frame without cacheability again. */
```

Comment:
I think this test should have a bit more explanation what is tested here and what not. If we map a frame uncached, revoke it and map it again uncached, the accesses always bypasses the cache and is supposed to go directly to memory. So all the test is dosing is checking that the kernel clears the content. What is does not test here is that proper cache management also happens, right? So we still need a test that maps a page cached, populated it with markers, then ensure this is written to RAM (this seems the tricky part). Then revoke the frame, map it again, but this time uncached. And then the whole frame should be empty, so we can be sure the page cleaning has not just happened in the cache, but it did not make it into RAM yet as no eviction has happened.

Reply:
Would not hurt to add that case for completeness, yes.

There are more, actually, there are also kernel objects, and as @kent-mcleod figured out last night also the i-cache. So ideally we'd add additional tests for:

1. cached framed with content -> retype -> uncached frame
2. other kernel objects with content -> retype -> uncached frame
3. any memory/cache content (any of the combinations from above) -> retype -> uncached frame mapped execute and trying to infer information from the i-cache

There's no really reliable way to make sure test 1 is written to RAM unless you manually flush it, which we could do for the test.

Reply:
It's possible to have a frame mapped in twice, once cached and once uncached. To test you hit RAM, you can write through the cached mapping and read back through the uncached mapping. Though this doesn't work on some of the common architectures where reads through uncached mappings still snoop the cache.

Reply:
> So all the test is dosing is checking that the kernel clears the content. What is does not test here is that proper cache management also happens, right?

The kernel doesn't clear the content through the same mapping, but through it's own separate mapping that is cached.  This test is checking that the clearing by the kernel ends up as ram.  It is checking that the proper cache management is occurring (for cleaning).  It doesn't check that other caches are invalidated.  If a future change was made such that the kernel reset memory via uncached mappings and didn't invalidate the caches, then this test wouldn't detect that a frame mapped cached after a retype could also see stale data.